### PR TITLE
improve: If `spokeRootsLookbackCount` is set, only try to execute root bundles from N most recent proposed root bundles rather than root bundle relays

### PR DESCRIPTION
**Pull request by nicholaspai**
Imagine we only want to execute leaves from the last 10 bundles. Intuitively this should only include leaves from the last 10 root bundles, but currently this means we count the latest 10 relayed root bundles for a chain. This means that some chains like Boba which have infrequent relayed root bundles, the lookback needs to be much larger.

This results in a lot of warnings: "Cannot validate bundle with insufficient event data. Set a larger DATAWORKER_FAST_LOOKBACK_COUNT". We should quiet these warnings

### CODE REVIEW

File: src/scripts/validateRootBundle.ts [hunk 1]
```diff
@@ -135,7 +135,8 @@ export async function validate(_logger: winston.Logger, baseSigner: Wallet): Pro
   // We want the bundle end of blocks following the target bundle so add +1 to the index.
   const overriddenConfig = {
     ...config,
-    dataworkerFastStartBundle: closestFollowingValidatedBundleIndex + 1,
+    dataworkerFastStartBundle:
+      closestFollowingValidatedBundleIndex === -1 ? "latest" : closestFollowingValidatedBundleIndex + 1,
```
Comment (nicholaspai) [inline src/scripts/validateRootBundle.ts:139]: Unrelated fix to this script if trying to validate bundle that is too recent and there no validated bundles after it

File: src/dataworker/Dataworker.ts [hunk 2]
```diff
@@ -800,18 +801,22 @@ export class Dataworker {
       message: "Executing slow relay leaves",
     });
 
+    let latestRootBundles = sortEventsDescending(this.clients.hubPoolClient.getValidatedRootBundles());
+    if (this.spokeRootsLookbackCount !== 0)
+      latestRootBundles = latestRootBundles.slice(0, this.spokeRootsLookbackCount);
```
Comment (james-a-morris) [inline src/dataworker/Dataworker.ts:806]: Nit: for readability, should we surround this statement with brackets?
```suggestion
    if (this.spokeRootsLookbackCount !== 0) {
      latestRootBundles = latestRootBundles.slice(0, this.spokeRootsLookbackCount);
    }
```

File: src/dataworker/Dataworker.ts [hunk 3]
```diff
@@ -1238,18 +1243,22 @@ export class Dataworker {
       message: "Executing relayer refund leaves",
     });
 
+    let latestRootBundles = sortEventsDescending(this.clients.hubPoolClient.getValidatedRootBundles());
+    if (this.spokeRootsLookbackCount !== 0)
+      latestRootBundles = latestRootBundles.slice(0, this.spokeRootsLookbackCount);
```
Comment (james-a-morris) [inline src/dataworker/Dataworker.ts:1248]: Nit: same as above

File: src/dataworker/Dataworker.ts [hunk 4]
```diff
@@ -1564,4 +1573,20 @@ export class Dataworker {
     if (leaf.groupIndex === 0) requiredAmount = requiredAmount.add(toBNWei("0.02"));
     return requiredAmount;
   }
+
+  // Filters out any root bundles that don't have a matching relayerRefundRoot+slowRelayRoot combination in the
+  // list of proposed root bundles `allRootBundleRelays`.
```
Comment (james-a-morris) [inline src/dataworker/Dataworker.ts:1578]: Nit: should we include valid JSDoc comments here so that IDE users have a full representation of this function?

File: src/dataworker/Dataworker.ts [hunk 5]
```diff
@@ -1564,4 +1573,20 @@ export class Dataworker {
...
+    rootBundles: ProposedRootBundle[],
+    allRootBundleRelays: RootBundleRelayWithBlock[]
+  ): RootBundleRelayWithBlock[] {
+    return allRootBundleRelays.filter(
+      (rootBundle) =>
+        rootBundles.find(
+          (_rootBundle) =>
+            _rootBundle.relayerRefundRoot === rootBundle.relayerRefundRoot &&
+            _rootBundle.slowRelayRoot === rootBundle.slowRelayRoot
+        ) !== undefined
```
Comment (james-a-morris) [inline src/dataworker/Dataworker.ts:1589]: Nit: semantically, would it make more sense to use `.some` instead of `.find` given that we're only interested in the boolean output?

**Review by james-a-morris — COMMENTED** [review]
Left some nits. This is a really interesting modification 🙂

**Review by james-a-morris — APPROVED** [review]
LGTM! Thanks for thoroughly reviewing my comments. Great work 😎

**Review by mrice32 — APPROVED** [review]
LGTM!
